tools/wptrunner/wptrunner/executors/base.py

```diff
@@ -515,45 +520,63 @@ def run_test(self, test):
             for i, node in enumerate(nodes):
                 success, data = self.get_hash(node, viewport_size, dpi, page_ranges)
                 if success is False:
-                    return {"status": data[0], "message": data[1]}
+                    test_results = {"status": data[0], "message": data[1]}
+                    if self.reftest_screenshot == "always" and log_data:
+                        test_results["extra"] = {"reftest_screenshots": log_data}
+                    return test_results
 
                 hashes[i], screenshots[i] = data
                 urls[i] = node.url
 
             is_pass, page_idx = self.check_pass(hashes, screenshots, urls, relation, fuzzy)
+
             if is_pass:
+                if self.reftest_screenshot == "always":
+                    log_data.extend(
+                        [{"url": urls[0], "screenshot": screenshots[0][page_idx],
+                          "hash": hashes[0][page_idx]},
+                         relation,
+                         {"url": urls[1], "screenshot": screenshots[1][page_idx],
+                          "hash": hashes[1][page_idx]}])
+
                 fuzzy = self.get_fuzzy(test, nodes, relation)
                 if nodes[1].references:
                     stack.extend(list(((nodes[1], item[0]), item[1])
                                       for item in reversed(nodes[1].references)))
                 else:
+                    test_result = {"status": "PASS", "message": None}
```

Comment:
Note that if the test is expected to `FAIL` then we should output screenshots if it actually get as `PASS` status and `reftest_screenshot` is set to `"unexpected"`.

Reply:
Done